044-cluster-wide-volume-usage-quota-management.md

```diff
@@ -0,0 +1,253 @@
+repetition is acceptable as our new limit types will be applied per-volume, so redundant volume descriptions don't
+impact the outcome.
+
+The Cluster Volume Source will use this API to discover volume information for the whole cluster.
+
+##### Admin Client Configuration
+
+To obtain log dir descriptions though the admin api we need to construct an admin client. We intend to
+connect to the replication listener and will piggyback on the broker's existing SSL configuration to
+obtain keystore/truststore properties.
+
+If using cluster sourced volumes we will expose these optional properties to customise the admin connection:
+
+1. `client.quota.callback.kafka.listener.name` (default `replication-9091`)
+2. `client.quota.callback.kafka.listener.port` (default `9091`)
+3. `client.quota.callback.kafka.listener.protocol` (default `SSL`)
+4. `client.quota.callback.kafka.clientIdPrefix` (default `__strimzi`)
+
+Setting `client.quota.callback.static.storage.hard` or `client.quota.callback.static.storage.soft` would be incompatible
+with the cluster sourced volumes and fail configuration of the plugin.
+
+The admin client will re-use the broker SSL configuration
+
+##### Limit Type Configuration
+
+When operating with cluster sourced volume data the existing limit types applied to the aggregate
+consumed bytes of all volumes would be meaningless. Nodes enter and exit the active set as part of normal operation.
+So we propose that the existing limiting method should be incompatible with cluster sourced volumes.
+
+Instead, we propose introducing new limit types applied on a per-volume basis. Meaning that there is a single value
+for each limit which we test against each volume. i.e. we do not support limiting based on a specific volume.
+
+We retain the existing concept of a soft limit (where throttling begins) and a hard limit (where we effectively
+stop message production) for all limit types.
+
+The only requirement is a single hard-limit configuration, if no soft-limit is configured we will infer that soft-limit==hard-limit.
+
+The different limit types can be combined, so you could combine a consumedBytesExceeds soft limit with a usableRatioBelow
+hard limit.
+
+Defining multiple soft limits or multiple hard limits would be an invalid state.
+
+The limits we want are:
+1. throttle if consumedBytes is greater-than-or-equal-to some threshold for any volume
+2. throttle if usableBytes is less-than-or-equal-to some threshold for any volume
+3. throttle if the ratio of usableBytes to capacityBytes is less-than-or-equal-to some threshold for any volume
+
+Using configuration properties like:
+
+1. `client.quota.callback.static.storage.perVolumeLimit.consumedBytesExceeds.soft=1000`
+2. `client.quota.callback.static.storage.perVolumeLimit.consumedBytesExceeds.hard=5000`
+3. `client.quota.callback.static.storage.perVolumeLimit.usableBytesBelow.soft=5000`
+4. `client.quota.callback.static.storage.perVolumeLimit.usableBytesBelow.hard=1000`
```

Comment:
So, in this example ... is it really 1000 and 5000 bytes? In that case, can you use something more realistic? Or is the limit name `bytes` but the actual unit different?

Reply:
Yep, we just put in arbitrary numbers. Have updated them to something more realistic.